commit f3809a18cf52480d169e8e33e4cb4de7a1bba7c5 (refs/changes/27/2727/3)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-10-06T16:24:33+02:00 (2 years ago)
    
    PAPI-143 Moray buckets need to be reindexed after schema updates

diff --git a/lib/backend.js b/lib/backend.js
index 2f68328..b4afaee 100644
--- a/lib/backend.js
+++ b/lib/backend.js
@@ -15,12 +15,12 @@
 
 var moray  = require('moray');
 var assert = require('assert');
-var util   = require('util');
 var jsprim = require('jsprim');
 var VError = require('verror');
 
 var RETRY_BUCKET_CREATION_INTERVAL = 10000; // in ms
 var DB_RESERVED_NAMES = ['default', 'group'];
+var MAX_ROWS_TO_REINDEX = 100;
 
 
 /*
@@ -60,6 +60,33 @@ function (cb) {
     var backend = createMorayClient(self.opts);
     self.backend = backend;
 
+    function reindexObjects(done) {
+        backend.reindexObjects(bucket, MAX_ROWS_TO_REINDEX,
+            function reindexCb(reindexErr, reindexRes) {
+            if (reindexErr) {
+                log.error({
+                    err: reindexErr,
+                    bucket: bucket
+                }, 'Reindex bucket error');
+                done(reindexErr);
+                return;
+            }
+
+            if (reindexRes.processed === 0) {
+                log.info('Reindex bucket done');
+                done();
+                return;
+            }
+
+            log.info({
+                bucket: bucket,
+                results: reindexRes
+            }, 'reindexed %d rows', reindexRes.processed);
+
+            reindexObjects(done);
+        });
+    }
+
     /*
      * Ensure bucket for packages in Moray exists. Keep retrying every 10s
      * until successful.
@@ -97,7 +124,7 @@ function (cb) {
                     morayConnectCallback();
                 }, RETRY_BUCKET_CREATION_INTERVAL);
             } else {
-                cb();
+                reindexObjects(cb);
             }
         });
     };
